lib/trip_dispatcher.rb

```diff
@@ -67,10 +61,119 @@ def inspect
+    # replace user objects with their respective driver objects in @passengers
+    def merge_drivers_and_users(drivers)
+      # find the drivers' user data and add their user info to their Driver object
+      drivers.each do |driver|
+        # Finding drivers user data based on id
+        user = self.find_passenger(driver.id)
+        # Adding user info to driver profile
+        if user
+          driver.name = user.name
+          driver.phone_number = user.phone_number
+          driver.trips = user.trips
+        end
+        # Swapping out passenger with driver
+        @passengers[@passengers.index(user)] = driver
+      end
+      return drivers
+    end
+
+    def load_drivers(filename)
+      # pull in drivers.csv and create Driver objects
+      drivers = load_drivers_from_csv(filename)
+      return merge_drivers_and_users(drivers) # drivers with user info
+    end
+
+    def load_driven_trips_to_drivers
+      trips.each do |trip|
+        trip_driver = find_driver(trip.driver.id)
+        trip_driver.add_driven_trip(trip)
+      end
+    end
 
     def check_id(id)
       raise ArgumentError, "ID cannot be blank or less than zero. (got #{id})" if id.nil? || id <= 0
     end
+
+    # creates and returns a new in progress trip
+    def create_new_trip(user_id, selected_driver)
+      # Creating new trip data
+      new_trip_data = {
+        id: (trips.last.id + 1),
+        passenger: find_passenger(user_id),
+        start_time: Time.now,
+        end_time: nil,
+        cost: nil,
+        rating: nil,
+        driver: selected_driver
+      }
+
+      # Creating a new instance of Trip
+      new_trip = Trip.new(new_trip_data)
+    end
+
+    # creates a new in progress trip, assigns an available driver
+    def request_trip(user_id)
+      # create a collection of only available drivers
+      available_drivers = @drivers.select { |driver| driver.status == :AVAILABLE }
+
+      # narrow the collection to drivers who have not driven before
+      available_drivers_no_trips = available_drivers.select {|driver| driver.driven_trips.length == 0}
+
+      # Ensure a driver who is a passenger is not assigned to themselves
+      if available_drivers.empty? ||
+        ( available_drivers.length == 1 && available_drivers[0].id == user_id )
+        raise RideShare::NoAvailableDriversError, "No drivers available at the moment"
+        # If the first available driver is not the user, assign the driver
+      elsif !available_drivers_no_trips.empty?
+        # Assign the first available driver with no existing driven trips
+        selected_driver = available_drivers_no_trips.first
+      else
+        # Pick the next available driver who has not driven in the longest time
+        available_drivers.sort_by! { |driver|
+          driver.driven_trips.max_by { |trip|
+            trip.end_time
+          }.end_time
+        }
+        selected_driver = available_drivers.first
+      end
```

Comment:
The process here is somewhat convoluted to outside eyes, and really, if this big `if` statement is either going to return a driver or raise an error, it makes sense to turn it into a helper method if only to help readers ignore the nitty gritty.